README.md
```
    _     _ _                 _          _
   (_)   | | |               | |        | |
    _  __| | |__   ___ ______| |__   ___| |_ __   ___ _ __
   | |/ _` | '_ \ / __|______| '_ \ / _ \ | '_ \ / _ \ '__|
   | | (_| | |_) | (__       | | | |  __/ | |_) |  __/ |
   | |\__,_|_.__/ \___|      |_| |_|\___|_| .__/ \___|_|
  _/ |                                    | |
 |__/                                     |_|
```

# jdbc-helper

A JDBC helper for Ruby/Database developers.
JDBCHelper::Connection object wraps around a JDBC connection and provides far nicer interface to
crucial database operations from primitive selects and updates to more complex ones involving
batch updates, prepared statements and transactions.
As the name implies, this gem only works on JRuby.

Tested on MySQL 5.5, Oracle 11g R2, PostgreSQL 9.0.4, MS SQL Server 2008 R2 and Cassandra 1.1.1 (CQL3).

## Installation
### Install gem
```
gem install jdbc-helper
```

### Setting up CLASSPATH
Add the appropriate JDBC drivers to the CLASSPATH.

```
export CLASSPATH=$CLASSPATH:~/lib/mysql-connector-java-5.1.16-bin.jar:~/lib/ojdbc6.jar
```

### In Ruby
```ruby
require 'jdbc-helper'
```

## Examples
### Connecting to a database
```ruby
# :driver and :url must be given
conn = JDBCHelper::Connection.new(
             driver: 'com.mysql.jdbc.Driver',
             url:    'jdbc:mysql://localhost/test')
conn.close


# Optional :user and :password
conn = JDBCHelper::Connection.new(
             driver:   'com.mysql.jdbc.Driver',
             url:      'jdbc:mysql://localhost/test',
             user:     'mysql',
             password: password)
conn.close
```

### Shortcut connectors

jdbc-helper provides shortcut connectors for the following databases
so that you don't have to specify lengthy class names and JDBC URLs.

- MySQL (`JDBCHelper::MySQL`)
- MariaDB (`JDBCHelper::MariaDB`)
- Oracle (`JDBCHelper::Oracle`)
- PostgreSQL (`JDBCHelper::PostgreSQL`)
- MS SQL Server (`JDBCHelper::MSSQL`)
- Cassandra (`JDBCHelper::Cassandra`)
- FileMaker Pro (`JDBCHelper::FileMaker`)
- SQLite (`JDBCHelper::SQLite`)

```ruby
mysql     = JDBCHelper::MySQL.connect(host, user, password, db)
mariadb   = JDBCHelper::MariaDB.connect(host, user, password, db)
oracle    = JDBCHelper::Oracle.connect(host, user, password, service_name)
postgres  = JDBCHelper::PostgreSQL.connect(host, user, password, db)
mssql     = JDBCHelper::MSSQL.connect(host, user, password, db)
cassandra = JDBCHelper::Cassandra.connect(host, keyspace)
filemaker = JDBCHelper::FileMaker.connect(host, user, password, db)
sqlite    = JDBCHelper::SQLite.connect(file_path)

# With extra parameters
mysql = JDBCHelper::MySQL.connect(host, user, password, db,
          rewriteBatchedStatements: true)

# With connection timeout of 30 seconds
mysql = JDBCHelper::MySQL.connect(host, user, password, db,
          rewriteBatchedStatements: true, timeout: 30)

# When block is given, connection is automatically closed after the block is executed
JDBCHelper::Cassandra.connect(host, keyspace) do |cc|
  # ...
end
```

### Querying database table

```ruby
conn.query('SELECT a, b, c FROM T') do |row|
  row.labels
  row.rownum

  a, b, c = row
  a, b, c = row.a,    row.b,    row.c    # Dot-notation
  a, b, c = row[0],   row[1],   row[2]   # Numeric index
  a, b, c = row['a'], row['b'], row['c'] # String index. Case-insensitive.
  a, b, c = row[:a],  row[:b],  row[:c]  # Symbol index. Case-insensitive.

  row[0..-1]                   # Range index. Returns an array of values.
  row[0, 3]                    # Offset and length. Returns an array of values.

  row.to_h                     # Row as a Hash
end

# You can even nest queries
conn.query('SELECT a FROM T') do |row1|
  conn.query("SELECT * FROM T_#{row1.a}") do |row2|
    # ...
  end
end

# Connection::ResultSet object is returned when block is not given
# - ResultSet is automatically closed when entirely iterated
rows = conn.query('SELECT * FROM T')
uniq_rows = rows.to_a.uniq

# However, partially consumed ResultSet objects *must be closed* manually
rset = conn.query('SELECT * FROM T')
rows = rset.take(2)
rset.close

# Enumerator chain
conn.query('SELECT * FROM LARGE_T').each_slice(1000).with_index do |slice, idx|
  slice.each do |row|
    # ...
  end
end
```

### Updating database table
```ruby
del_count = conn.update('DELETE FROM T')
```

### Executing any SQL
```ruby
rset = conn.execute('SELECT * FROM T')
rset.each do |row|
  # Returned result must be used or closed
end

del_count = conn.execute('DELETE FROM T')
```

### Transaction
```ruby
committed = conn.transaction do |tx|
  # ...
  # Transaction logic here
  # ...

  if success
    tx.commit
  else
    tx.rollback
  end
  # You never reach here.
end
```

### Using batch interface
```ruby
conn.add_batch('DELETE FROM T')
conn.execute_batch
conn.add_batch('DELETE FROM T')
conn.clear_batch
```

### Using prepared statements
```ruby
p_sel = conn.prepare('SELECT * FROM T WHERE b = ? and c = ?')
p_sel.query(100, 200) do |row|
  p row
end
p_sel.close

p_upd = conn.prepare('UPDATE T SET a = ? WHERE b = ?')
count = 0
100.times do |i|
  count += p_upd.update('updated a', i)
end

p_upd.add_batch('pstmt + batch', 10)
p_upd.add_batch('pstmt + batch', 20)
p_upd.add_batch('pstmt + batch', 30)
p_upd.execute_batch
p_upd.close
```

### Accessing underlying Java object with `java` method

```ruby
conn.java.setAutoCommit false

pstmt = conn.prepare(sql)
pstmt.java.getMetaData
```

### Using table wrappers (since 0.2.0)
```ruby
# Creates a table wrapper
table = conn.table('test.data')
# Or equievalently,
table = conn['test.data']

# Counting the records in the table
table.count
table.count(a: 10)
table.where(a: 10).count

table.empty?
table.where(a: 10).empty?

# Selects the table by combining select, where, order, limit and fetch_size methods
table.select('a apple', :b).where(c: (1..10)).order('b desc', 'a asc').fetch_size(100).limit(1000).each do |row|
  puts row.apple
end

# Build select SQL
sql = table.select('a apple', :b).where(c: (1..10)).order('b desc', 'a asc').sql

# Updates with conditions
table.where(c: 3).update(a: 'hello', b: { sql: 'now()' })

# Insert into the table
table.insert(a: 10, b: 20, c: { sql: '10 + 20' })
table.insert_ignore(a: 10, b: 20, c: 30)
table.replace(a: 10, b: 20, c: 30)

# Update with common default values
with_defaults = table.default(a: 10, b: 20)
with_defaults.insert(c: 30)
with_defaults.where('a != 10 or b != 20').update   # sets a => 10, b => 20

# Batch updates with batch method
table.batch.insert(a: 10, b: 20, c: { sql: '10 + 20' })
table.batch.insert_ignore(a: 10, b: 20, c: 30)
table.batch.where(a: 10).update(a: 20)
table.execute_batch :insert, :update

# Delete with conditions
table.delete(c: 3)
# Or equivalently,
table.where(c: 3).delete

# Truncate or drop table (Cannot be undone)
table.truncate!
table.drop!
```

#### Building complex where clauses
```ruby
# With any number of Strings, Arrays and Hashes
scope = table.where(
  "x <> 'hello'",                # x <> 'hello'
  ["y = ? or z > ?", 'abc', 10], # and (y = 'abc' or z > 10)
  a: 'abc',                      # and a = 'abc'
  b: (1..10),                    # and b between 1 and 10
  c: (1...10),                   # and c >= 1 and c < 10
  d: %w[a b c],                  # and d in ('a', 'b', 'c')
  e: { sql: 'sysdate' },         # and e = sysdate
  f: { not: nil },               # and f is not null
  g: { gt: 100, le: 200 },       # and g > 100 and g <= 200
  h: { lt: 100 },                # and h < 100
  i: { like: 'ABC%' },           # and i like 'ABC%'
  j: { not: { like: 'ABC%' } },  # and j not like 'ABC%'
  k: { le: { sql: 'sysdate' } }  # and k <= sysdate
)
scope.update(a: 'xyz')
```

#### Invalid use of plain String conditions

A TableWrapper object internally builds SQL strings
and creates JDBC PreparedStatement object for each distinct SQL.

If you build many number of where-clause Strings as shown in the following code,
soon there will be too many open PreparedStatements,
and if the number exceeds the system limit, an error will be thrown.

```ruby
table = connection['table']

# Leads to 10000 PreparedStatements !!
10000.times do |idx|
  table.count("id = #{idx}")
    # select count(*) from table where id = 0
    # select count(*) from table where id = 1
    # select count(*) from table where id = 2
    # select count(*) from table where id = 3
    # ...
end
```

In that case, you can `close` the table wrapper to close all the open PreparedStatements.

```ruby
table.close
```

However, you should always prefer using much more efficient Hash or Array expression over plain String,
so you don't have to worry about the proliferation of PreparedStatements.

```ruby
# 20000 queries but only a single PreparedStatement
10000.times do |idx|
  # 1. with Hash
  table.count('id' => idx)
    # select count(*) from table where id = ?

  # 2. with Array
  table.count(["id = ?", idx])
    # select count(*) from table where id = ?
end
```

### Using function wrappers (since 0.2.2)
```ruby
conn.function(:mod).call 5, 3
conn.function(:coalesce).call(nil, nil, 'king')
```

### Using procedure wrappers (since 0.3.0)
```ruby
# Working with IN/INOUT/OUT parameteres
# Bind by ordinal number
conn.procedure(:update_and_fetch_something).call(
         100,                 # Input parameter
         ["value", String],   # Input/Output parameter
         Fixnum               # Output parameter
)

# Bind by parameter name
conn.procedure(:update_and_fetch_something).call(
         a: 100, b: ["value", String], c: Fixnum)
```

### Using sequence wrappers (since 0.4.2)
```ruby
seq = conn.sequence(:my_seq)
next = seq.nextval
curr = seq.currval
seq.reset!
seq.reset! 100
```

## Contributing to jdbc-helper

* Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet
* Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it
* Fork the project
* Start a feature/bugfix branch
* Commit and push until you are happy with your contribution
* Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.
* Please try not to mess with the Rakefile, version, or history. If you want to have your own version, or is otherwise necessary, that is fine, but please isolate to its own commit so I can cherry-pick around it.

## Contributors

* [Larry Staton Jr.](https://github.com/statonjr)

## Copyright

Copyright (c) 2011 Junegunn Choi. See LICENSE.txt for
further details.


CHANGELOG.md
### 0.8.3

* Fixed not to ignore property with false value

### 0.8.2

* Allow :driver parameter to be an object that responds to `connect`

### 0.8.1

* Allow :driver parameter to be in either String or Class

### 0.8.0

0.8.0 introduces a few backward-incompatible changes.

* `Connection#ResultSetEnumerator` is renamed to `Connection::ResultSet`
* `Connection#query` method will return a ResultSet object instead of an Array
  * `Connection#enumerate` method is now retired, and just a synonym for query method
  * Partially consumed ResultSet must be closed explicitly
* `ResultSet#each` method will return an enumerator when block is not given
* Refined TableWrapper interface with external [sql_helper](https://github.com/junegunn/sql_helper) gem
  * The use of `JDBCHelper::SQL` is deprecated
* Added MariaDB connector
* Added SQLite connector

### 0.7.7 / 2013/01/0?
* `PreparedStatment`s and `TableWrapper`s now inherit the fetch size of the connection
* Added `JDBCHelper::TableWrapper#fetch_size`
* Added `JDBCHelper::TableWrapper#execute_batch`
* Added `JDBCHelper::TableWrapper#clear_batch`
* `execute_batch` method returns the sum of all update counts
* Removed undocumented operation statistics

### 0.7.6 / 2012/08/26
* Added FileMaker Pro connector (statonjr)

### 0.7.5 / 2012/07/28
* `require 'jdbc-helper/sql'` is allowed in non-Java environments

### 0.7.4 / 2012/07/09
* Revised Connectors
 * Use `JDBCHelper::MySQL` instead of `JDBCHelper::MySQLConnector`
  * Optional `timeout` parameter has been removed. For backward-compatibility, `XXXConnector` still receives timeout parameter, though deprecated.
* Added Cassandra connector
 * `JDBCHelper::Cassandra#connect`

### 0.7.3 / 2012/05/11
* Fixed occasional error when rolling back transaction: "Can't call rollback when autocommit=true"

### 0.7.2 / 2012/03/07
* Added JDBCHelper::Connection#execute, JDBCHelper::PreparedStatement#execute
 * Returns update count as Fixnum or JDBCHelper::Connection::ResultSetEnumerator
   depending on the type of the given statement.
* Bug fix: Setting nil value for PreparedStatements fails on SQL Server
 * Modified to refer to ParameterMetaData when calling setNull method
 * Current version of SQL Server driver has some crazy bugs with PreparedStatements.
   You may notice that SQL Server complains on a perfectly fine SQL statement
   only when you try to access the metadata of its wrapping PreparedStatement.
   For such cases, where we can't refer to ParameterMetaData (and you will see a warning log),
   setting nulls should still fail.

### 0.7.1 / 2012/02/28
* Modified nextval/currval syntax in JDBCHelper::SequenceWrapper
 * PostgreSQL: nextval('[sequence]')
 * Others: [sequence].nextval
* Bug fix: JDBCHelper::Connection#inspect returns invalid data

### 0.7.0 / 2012/01/30
* Helper methods for generating where clauses for TableWarppers
 * `JDBCHelper::SQL.expr` (`JDBCHelper::SQL` deprecated in favor of `expr`)
 * `JDBCHelper::SQL.ne`
 * `JDBCHelper::SQL.gt`
 * `JDBCHelper::SQL.ge`
 * `JDBCHelper::SQL.lt`
 * `JDBCHelper::SQL.le`
 * `JDBCHelper::SQL.like`
 * `JDBCHelper::SQL.not_like`
 * `JDBCHelper::SQL.not_null`
* Tested on PostgreSQL
* Tested on MS SQL Server
* JDBCHelper::PostgresConnector added.
* JDBCHelper::SqlServerConnector added.


### 0.6.3 / 2011/11/09
* Added flexibility when writing TableWrapper conditions with the new ActiveRecord-like Array-expression.

```ruby
table.count(["title != ? or year <= ?", 'N/A', '2011'], :month => (1..3), :day => 10)
```

### 0.6.2 / 2011/10/27
* Bug fix: JDBCHelper::Connection#initialize *still* modifies its input Hash

### 0.6.1 / 2011/09/05
* Minor fix for an inefficient type conversion
* More tests. Test coverage over 97%.

### 0.6.0 / 2011/09/05
* Datatype handling
 * Proper BLOB handling.
 * Proper handling of Numeric types.
  * Numeric value will be returned as Fixnum, Bignum, Float or BigDecimal. (No more Strings!)
 * ParameterizedStatement#set_param achieves better precision with Time#to_f
* Increased test coverage

### 0.5.1 / 2011/08/29
* Bug fix: TableWrapper objects are cached to prevent proliferation of PreparedStatements.

```ruby
# In the previous version, codes like the following generated too many prepared statements.
(0..10000).each do |i|
  db.table(:my_table).count(:id => i)
end
```

### 0.5.0 / 2011/08/29
* TableWrapper internally uses PreparedStatements instead of regular statements for better performance. <b>However, this makes it impossible to serialize batched operations.</b> Thus, you should not use batched statements with TableWrapper when the order of operations is important.
* TableWrapper#close closes cached PreparedStatements.
* Connection#execute_batch and Connection#clear_batch have been modified to execute or clear every PreparedStatment of the Connection.
* Connection overrides clone method to create another connection to the same database.

### 0.4.10 / 2011/08/09
* Sadly, I found that with Oracle, precision and scale information from ResultSetMetaData is completely unreliable. Going conservative.

### 0.4.9 / 2011/08/09
* Oracle NUMBER datatype is retrived as Fixnum or Bignum or String depending on its precision and scale.

### 0.4.8 / 2011/07/12
* Improved fetch size interface
* Supports non-string property values in the input parameter for Connection object

### 0.4.7 / 2011/06/28
* PreparedStatement performance improvement. Minimized unneccesary to_s call when setting parameters. 10% to 30% improvement observed for usual bulk insertion.

### 0.4.6 / 2011/06/20
* Bug fix: Invalid alias for JDBCHelper::TableWrapper#drop!
* Bug fix: JDBCHelper::Connection#initialize modifies its input Hash
* Added attr_readers for url and driver to JDBCHelper::Connection

### 0.4.5 / 2011/06/10
* JDBCHelper::TableWrapper#default method added which allows you to specify common default values for the subsequent inserts and updates.

### 0.4.4 / 2011/06/08

lib/jdbc-helper/connection.rb
# encoding: UTF-8
# Junegunn Choi (junegunn.c@gmail.com)

require 'insensitive_hash/minimal'

require 'jdbc-helper/connection/type_map'
require 'jdbc-helper/connection/parameterized_statement'
require 'jdbc-helper/connection/prepared_statement'
require 'jdbc-helper/connection/callable_statement'
require 'jdbc-helper/connection/statement_pool'
require 'jdbc-helper/connection/result_set'
require 'jdbc-helper/connection/row'

require 'jdbc-helper/wrapper/object_wrapper'
require 'jdbc-helper/wrapper/table_wrapper'
require 'jdbc-helper/wrapper/table_wrapper'
require 'jdbc-helper/wrapper/sequence_wrapper'
require 'jdbc-helper/wrapper/function_wrapper'
require 'jdbc-helper/wrapper/procedure_wrapper'

module JDBCHelper
# Encapsulates JDBC database connection.
# Lets you easily execute SQL statements and access their results.
#
# @example Prerequisites
#  # Add JDBC driver of the DBMS you're willing to use to your CLASSPATH
#  export CLASSPATH=$CLASSPATH:~/lib/mysql-connector-java.jar
#
#
# @example Connecting to a database
#
#  # :driver and :url must be given
#  conn = JDBCHelper::Connection.new(
#               :driver => 'com.mysql.jdbc.Driver',
#               :url    => 'jdbc:mysql://localhost/test')
#  conn.close
#
#
#  # Optional :user and :password
#  conn = JDBCHelper::Connection.new(
#               :driver   => 'com.mysql.jdbc.Driver',
#               :url      => 'jdbc:mysql://localhost/test',
#               :user     => 'mysql',
#               :password => '')
#  conn.close
#
#
#  # MySQL shortcut connector
#  conn = JDBCHelper::MySQLConnector.connect('localhost', 'mysql', '', 'test')
#  conn.close
#
# @example Querying database table
#
#  conn.query("SELECT a, b, c FROM T") do | row |
#      p row.labels
#      p row.rownum
#
#      puts row.a, row.b, row.c
#      puts row[0], row[1], row[2]
#      puts row['a'], row['b'], row['c']
#  end
#
#  # Returns an array of rows when block is not given
#  rows = conn.query("SELECT b FROM T")
#  uniq_rows = rows.uniq
#
#  # You can even nest queries
#  conn.query("SELECT a FROM T") do | row1 |
#      conn.query("SELECT * FROM T_#{row1.a}") do | row2 |
#          # ...
#      end
#  end
# @example Updating database table
#  del_count = conn.update("DELETE FROM T")
#
# @example Transaction
#  committed = conn.transaction do | tx |
#      # ...
#      # Transaction logic here
#      # ...
#
#      if success
#          tx.commit
#      else
#          tx.rollback
#      end
#  end
#
# @example Using batch interface
#  conn.add_batch("DELETE FROM T");
#  conn.execute_batch
#  conn.add_batch("DELETE FROM T");
#  conn.clear_batch
#
# @example Using prepared statements
#  p_sel = conn.prepare("SELECT * FROM T WHERE b = ? and c = ?")
#  p_sel.query(100, 200) do | row |
#      p row
#  end
#  p_sel.close
#
#  p_upd = conn.prepare("UPDATE T SET a = ? WHERE b = ?")
#  count = 0
#  100.times do | i |
#      count += p_upd.update('updated a', i)
#  end
#
#  p_upd.add_batch('pstmt + batch', 10)
#  p_upd.add_batch('pstmt + batch', 20)
#  p_upd.add_batch('pstmt + batch', 30)
#  p_upd.execute_batch
#  p_upd.close
class Connection
  # JDBC URL of the connection
  # @return [String]
  attr_reader :url

  # JDBC driver of the connection
  # @return [String|Class|#connect]
  attr_reader :driver

  # Returns the underlying JDBC Connection object.
  # Only use this when you really need to access it directly.
  def jdbc_conn
    @conn
  end
  alias java_obj jdbc_conn
  alias java     jdbc_conn

  # Creates a database connection.
  # - `args` hash must include :driver (or "driver") and :url (or "url")
  # - and takes optional :user and :password tuples (or "user", "password")
  # - You can also specify :timeout (or "timeout") to override the default connection timeout (60 seconds)
  #
  # Must be closed explicitly if not used.
  # If a block is given, the connection is automatically closed after executing the block.
  # @param [Hash] args
  def initialize(args = {})
    # Subsequent deletes should not affect the input
    @args = args
    args = InsensitiveHash[ @args ]

    raise ArgumentError.new("driver not given") unless args.has_key? :driver
    raise ArgumentError.new("url not given") unless args.has_key? :url

    @driver = args.delete :driver
    @url = args.delete :url

    timeout = args.has_key?(:timeout) ? args.delete(:timeout) : Constants::DEFAULT_LOGIN_TIMEOUT
    if timeout
      if timeout.is_a?(Fixnum) == false || timeout <= 0
        raise ArgumentError.new("Timeout must be a positive integer")
      end
      Java::java.sql.DriverManager.setLoginTimeout timeout
    end

    props = Java::java.util.Properties.new
    args.each do |k, v|
      props.setProperty(k.to_s, v.to_s) unless v.nil?
    end

    @conn = case @driver
            when String
              # NameError will be thrown for invalid drivers
              Java::JavaClass.for_name @driver
              Java::java.sql.DriverManager.get_connection(@url, props)
            when Class
              @driver.new.connect(@url, props)
            else
              if @driver.respond_to?(:connect)
                @driver.connect(@url, props)
              else
                raise ArgumentError.new('Invalid type for :driver')
              end
            end
    @spool = StatementPool.send :new, self
    @bstmt = nil
    @fetch_size = nil

    @pstmts = []

    @table_wrappers = {}

    if block_given?
      begin
        yield self
      ensure
        close rescue nil
      end
    end
  end

  # Creates another connection with the same parameters as this Connection.
  # @return [JDBCHelper::Connection]
  def clone
    nc = JDBCHelper::Connection.new @args
    nc.fetch_size = @fetch_size if @fetch_size
    nc
  end

  # Creates a prepared statement, which is also an encapsulation of Java PreparedStatement object
  # @param [String] qstr SQL string
  def prepare(qstr)
    check_closed

    pstmt = PreparedStatement.send(:new, self, qstr, @conn.prepare_statement(qstr))
    pstmt.set_fetch_size @fetch_size if @fetch_size

    @pstmts << pstmt
    pstmt
  end

  # @return [Array] Prepared statements currently opened for this connection
  def prepared_statements
    @pstmts
  end

  # Creates a callable statement.
  # @param [String] qstr SQL string
  def prepare_call(qstr)
    check_closed

    CallableStatement.send(:new, self, qstr, @conn.prepare_call(qstr))
  end

  # Executes the given code block as a transaction. Returns true if the transaction is committed.
  # A transaction object is passed to the block, which only has commit and rollback methods.
  # The execution breaks out of the code block when either of the methods is called.
  # @yield [JDBCHelper::Connection::Transaction] Responds to commit and rollback.
  # @return [Boolean] True if committed
  def transaction
    check_closed

    raise ArgumentError.new("Transaction block not given") unless block_given?
    tx = Transaction.send :new, @conn
    ac = @conn.get_auto_commit
    status = :unknown
    begin
      @conn.set_auto_commit false
      yield tx
      @conn.commit
      status = :committed
    rescue Transaction::Commit
      status = :committed
    rescue Transaction::Rollback
      status = :rolledback
    ensure
      @conn.rollback if status == :unknown && @conn.get_auto_commit == false
      @conn.set_auto_commit ac
    end
    status == :committed
  end

  # Executes an SQL and returns the count of the update rows or a ResultSet object
  # depending on the type of the given statement.
  # If a ResultSet is returned, it must be enumerated or closed.
  # @param [String] qstr SQL string
  # @return [Fixnum|ResultSet]
  def execute(qstr)
    check_closed

    stmt = @spool.take
    begin
      if stmt.execute(qstr)
        ResultSet.send(:new, stmt.getResultSet) { @spool.give stmt }
      else
        rset = stmt.getUpdateCount
        @spool.give stmt
        rset
      end
    rescue Exception => e
      @spool.give stmt
      raise
    end
  end

  # Executes an update and returns the count of the updated rows.
  # @param [String] qstr SQL string
  # @return [Fixnum] Count of affected records
  def update(qstr)
    check_closed

    @spool.with do | stmt |
      ret = stmt.execute_update(qstr)
    end
  end

  # Executes a select query.
  # When a code block is given, each row of the result is passed to the block one by one.
  # If not given, ResultSet is returned, which can be used to enumerate through the result set.
  # ResultSet is closed automatically when all the rows in the result set is consumed.
  #
  # @example Nested querying
  #   conn.query("SELECT a FROM T") do | trow |
  #     conn.query("SELECT * FROM U_#{trow.a}").each_slice(10) do | urows |
  #       # ...
  #     end
  #   end
  # @param [String] qstr SQL string
  # @yield [JDBCHelper::Connection::Row]
  # @return [Array]
  def query(qstr, &blk)
    check_closed

    stmt = @spool.take
    begin
      rset = stmt.execute_query(qstr)
    rescue Exception => e
      @spool.give stmt
      raise
    end

    enum = ResultSet.send(:new, rset) { @spool.give stmt }
    if block_given?
      enum.each do |row|
        yield row
      end
    else
      enum
    end
  end
  alias enumerate query

  # Adds a statement to be executed in batch
  # Adds to the batch
  # @param [String] qstr
  # @return [NilClass]
  def add_batch(qstr)
    check_closed

    @bstmt ||= @spool.take
    @bstmt.add_batch qstr
  end

  # Executes batched statements including prepared statements. No effect when no statement is added
  # @return [Fixnum] Sum of all update counts
  def execute_batch
    check_closed

    cnt = 0

    if @bstmt
      cnt += @bstmt.execute_batch.inject(:+) || 0
      @spool.give @bstmt
      @bstmt = nil
    end

    @pstmts.each do |pstmt|
      cnt += pstmt.execute_batch
    end

    cnt
  end

  # Clears the batched statements including prepared statements.
  # @return [NilClass]
  def clear_batch
    check_closed

    if @bstmt
      @bstmt.clear_batch
      @spool.give @bstmt
      @bstmt = nil
    end

    @pstmts.each do |stmt|
      stmt.clear_batch
    end
  end

  # Gives the JDBC driver a hint of the number of rows to fetch from the database by a single interaction.
  # This is only a hint. It may have no effect at all.
  # @param [Fixnum] fsz
  # @return [NilClass]
  def set_fetch_size(fsz)
    check_closed

    @fetch_size = fsz
    @spool.each { | stmt | stmt.set_fetch_size @fetch_size }
  end
  alias fetch_size= set_fetch_size

  # Returns the fetch size of the connection. If not set, nil is returned.
  # @return [Fixnum]
  attr_reader :fetch_size

  # Closes the connection
  # @return [NilClass]
  def close
    return if closed?
    @spool.close
    @conn.close
    @conn = @spool = nil
  end

  # Returns if this connection is closed or not
  # @return [Boolean]
  def closed?
    @conn.nil?
  end

  # Returns a table wrapper for the given table name
  # @since 0.2.0
  # @param [String/Symbol] table_name Name of the table to be wrapped
  # @return [JDBCHelper::TableWrapper]
  def table table_name
    table = JDBCHelper::TableWrapper.new(self, table_name)
    table = table.fetch_size(@fetch_size) if @fetch_size
    @table_wrappers[table_name] ||= table
  end
  alias [] table

  # Returns a sequence wrapper for the given name
  # @since 0.4.2
  # @param [String/Symbol] sequence_name Name of the sequence to be wrapped
  # @return [JDBCHelper::SequenceWrapper]
  def sequence sequence_name
    JDBCHelper::SequenceWrapper.new self, sequence_name
  end

  # Returns a function wrapper for the given function name
  # @since 0.2.2
  # @param [String/Symbol] func_name Name of the function to be wrapped
  # @return [JDBCHelper::FunctionWrapper]
  def function func_name
    JDBCHelper::FunctionWrapper.new self, func_name
  end

  # Returns a procedure wrapper for the given procedure name
  # @since 0.3.0
  # @param [String/Symbol] proc_name Name of the procedure to be wrapped
  # @return [JDBCHelper::ProcedureWrapper]
  def procedure proc_name
    JDBCHelper::ProcedureWrapper.new self, proc_name
  end

  # @return [String]
  def inspect
    InsensitiveHash[@args].merge({ :closed? => closed? }).tap { |c|
      c.delete(:password)
    }.inspect
  end

private
  # Transaction object passed to the code block given to transaction method
  class Transaction
    # Commits the transaction
    # @raise [JDBCHelper::Transaction::Commit]
    def commit
      @conn.commit
      raise Commit
    end
    # Rolls back this transaction
    # @raise [JDBCHelper::Transaction::Rollback]
    def rollback
      @conn.rollback
      raise Rollback
    end
  private
    def initialize(conn) # :nodoc:
      @conn = conn
    end
    class Commit < Exception # :nodoc:
    end
    class Rollback < Exception # :nodoc:
    end
  end

  def create_statement # :nodoc:
    stmt = @conn.create_statement
    stmt.set_fetch_size @fetch_size if @fetch_size
    stmt
  end

  def close_pstmt pstmt
    @pstmts.delete pstmt
  end

  def check_closed
    raise RuntimeError.new('Connection already closed') if closed?
  end
end#Connection
end#JDBCHelper

